### PR TITLE
Marekhorst 1475 rewrite citation matching algorithm in spark 2.4

### PR DESCRIPTION
**Pull request by marekhorst**
This PR includes all the changes required to rewrite citationmatching code from spark 1.6 to spark 2.4 (and from scala 2.10 to scala 2.11).

I decided to introduce a new `iis-3rdparty-coansys-citation-matching` module encapsulating the whole code moved from CoAnSys (see `README.markdown` file for details) and adjusted to spark2.4/scala2.11 while having `iis-wf-citationmatching` updated in just 3-4 places in the java code (mostly related to `Iterator`s) and having `pom.xml` dependencies updated. The code moved to `iis-3rdparty-coansys-citation-matching` was sourced from `citation-matching-core-code` and `ceon-scala-commons-lite` CoAnSys modules. `ceon-scala-commons-lite` was required because it included scala classes utilized by `citation-matching-core-code` and I had to have them recompiled with scala 2.11 in order to make citation matching spark2.4 compliant.

This pull request includes several WiP commits, most of them are bound to #1475 but one of them is related to #1481 (introduction of two spark timeout related properties). Once I get to the point of merging with the master branch I will either squash those commits to the one with referencing two issues in the commit message or split into two separate commits.

### CODE REVIEW

**Comment by marekhorst** [comment]
I found this useful hint (`Alt + click` on any chevron icon in the file’s header):
https://dzhavat.github.io/2021/11/08/quick-way-to-collapse-expand-all-files-at-once-on-a-pull-request-on-github.html
while dealing with such a great number of changed files e.g. to focus only on a subset of changes related to CoAnSys or IIS codebase.

File: iis-3rdparty-coansys-citation-matching/README.markdown [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# About
+
+All the scala, java and resource files in this module stored within the pl/edu/icm/ subdirectories were copied from the following CoAnSys repository locations:
+
+https://github.com/CeON/CoAnSys/tree/master/citation-matching/citation-matching-coansys-code
```
Comment (mpol) [inline iis-3rdparty-coansys-citation-matching/README.markdown:5]: It is actually from `citation-matching-core-code`

File: iis-3rdparty-coansys-citation-matching/README.markdown [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+https://github.com/CeON/CoAnSys/tree/master/ceon-scala-commons-lite
```
Comment (mpol) [inline iis-3rdparty-coansys-citation-matching/README.markdown:7]: I think we should document the exact commit `f0373f0416c93d3cff3a6ea137da09fba1481cad` from which the files were copied.

**Review by mpol — CHANGES_REQUESTED** [review]
Everything looks OK regarding the code, but the new README needs some fixes.

**Comment by marekhorst** [comment]
@mpol I have just refined the `README.markdown` file content a little bit.

**Review by mpol — APPROVED** [review]
LGTM

**Comment by marekhorst** [comment]
Changes covered with this PR were merged with this commit:

38d5290ae84ca732b786ce49fc16db7d99367bb3

after rebasing and squashing all the commits into the one.

**Pull request closed without merging**